File: Day-05/rust/emmaboecker/day05.rs
use std::cmp::Ordering;

pub fn main() {
    let input = include_str!("day05.txt");

    let (rules, pages) = input.split_at(input.find("\r\n\r\n").unwrap());

    let rules = rules
        .lines()
        .map(|rule| {
            let (before, after) = rule.split_once("|").unwrap();
            (
                before.trim().parse::<i32>().unwrap(),
                after.trim().parse::<i32>().unwrap(),
            )
        })
        .collect::<Vec<_>>();

    let page_lines = pages.lines().skip(2).map(|line| {
        line.split(",")
            .map(|page| page.parse::<i32>().unwrap())
            .collect::<Vec<_>>()
    });

    let mut part1 = 0;
    let mut part2 = 0;
    for pages in page_lines {
        let mut is_sorted = true;
        'page: for (i, page) in pages.iter().enumerate() {
            for (before, after) in rules
                .iter()
                .filter(|(before, after)| pages.contains(before) && pages.contains(after))
            {
                if after == page && !pages[..i].iter().any(|p| p == before) {
                    is_sorted = false;
                    break 'page;
                }
            }
        }
        if is_sorted {
            part1 += pages[pages.len() / 2];
        } else {
            let mut sorted = pages.clone();
            sorted.sort_by(|a, b| {
                let rule = rules
                    .iter()
                    .find(|(before, after)| a == before && b == after);
                if let Some((before, after)) = rule {
                    if a == before && b == after {
                        Ordering::Less
                    } else {
                        Ordering::Greater
                    }
                } else {
                    Ordering::Equal
                }
            });
            part2 += sorted[sorted.len() / 2];
        }
    }

    println!("Part 1: {:?}", part1);
    println!("Part 2: {:?}", part2);
}

File: Day-05/kotlin/maxbossing/Day05.kt
import java.nio.file.Path

class Day05 {
    private val input = Path("day05.txt").readText()

    val rules = input
        .split(Regex("\\R\\s*\\R"))
        .toList()[0]
        .split("\n")
        .map { it.trim() }
        .filter { it.isNotBlank() }
        .map { it.split("|") }
        .map { it[0].trim().toInt() to it[1].trim().toInt() }

    val books = input
        .split(Regex("\\R\\s*\\R"))
        .toList()[1]
        .split("\n")
        .map { it.trim() }
        .filter { it.isNotBlank() }
        .map { it.split(",").map { it.toInt() } }

    fun part1() = books.filter { pages ->
        rules.all { (before, after) ->
            if (before !in pages || after !in pages) return@all true
            return@all pages.indexOf(before) < pages.indexOf(after)
        }
    }.sumOf { it[it.size / 2] }

    fun part2(): Int = TODO()

}